### PR TITLE
fix(ssa): conditionalise array indexes under IF statements

### PR DESCRIPTION
**Pull request by guipublic**
# Description

## Problem\*

Issue reported by Nouns here: https://github.com/aragonzkresearch/noir-storage-proofs/blob/14cb94207fc466e4682ac2717555e5a7dd229f4a/lib/src/lib.nr#LL244C2-L251C13

Same error is also reported here: https://github.com/noir-lang/noir/issues/1347

## Summary\*

This PR sets out to conditionalise array access indexes that are under if statement to avoid out-of-bounds errors.

### Example

cf. regression test added in array_dynamic

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [X] I have tested the changes locally.
- [X] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: crates/nargo_cli/tests/test_data/array_dynamic/src/main.nr [hunk 1]
```diff
@@ -1,24 +1,28 @@
...
 }
 
 fn dyn_array(mut x: [u32; 5], y: Field, z: Field) {
-    assert(x[y] == 111);
-    assert(x[z] == 101);
+   assert(x[y] == 111);
+   assert(x[z] == 101);
     x[z] = 0;
-    assert(x[y] == 111);
-    assert(x[1] == 0);
```
Comment (jfecher) [inline crates/nargo_cli/tests/test_data/array_dynamic/src/main.nr:21]: Looks like the spacing got messed up here

File: crates/noirc_evaluator/src/ssa/conditional.rs [hunk 2]
```diff
@@ -577,6 +577,35 @@ impl DecisionTree {
...
+                        ));
+                        optimizations::simplify_id(ctx, idx).unwrap();
+                        stack.push(idx);
+
+                        let ins2 = ctx.instruction_mut(ins_id);
+                        ins2.operation = Operation::Load {
+                            array_id: *array_id,
+                            index: idx,
+                            location: *location,
+                        };
```
Comment (jfecher) [inline crates/noirc_evaluator/src/ssa/conditional.rs:608]: So a conditional load `if c { a[i] }` is translated essentially to `a[c * i]` if I'm reading this correctly. Makes sense, but it is not obvious why we're doing it from the code alone so lets document it.

File: crates/nargo_cli/tests/test_data/array_dynamic/src/main.nr (hunk 1 above)
Comment (guipublic) [reply crates/nargo_cli/tests/test_data/array_dynamic/src/main.nr]: Done

File: crates/noirc_evaluator/src/ssa/conditional.rs (hunk 2 above)
Comment (guipublic) [reply crates/noirc_evaluator/src/ssa/conditional.rs:609]: I have added a comment, and also added a test to do it only when the index is not constant.

**Review by jfecher — APPROVED** [review]

**Pull request merged**